relation_engine_server/test/test_json_validation.py

```diff
@@ -315,6 +358,74 @@ def date_format_validation(self, schema_arg, schema_file_arg):
                 data_file=file_path,
                 validate_at=valid_json_loc)
 
+    def test_array_validation(self, schema_arg=None, schema_file_arg=None):
```

Comment:
Why are we testing this? It makes sense to me to test something like setting defaults, since that involves our own custom code. But array validation is basic functionality covered by a third party lib.

Reply:
I had a weird problem with arrays not validating properly, and I couldn't find anything in the `python-jsonschema` docs or the generic `jsonschema` docs to explain what was going on. I wrote some tests to work out why things weren't working, and then edited the test input until the tests passed. Between `python-jsonschema` and `pyyaml`, I don't have much trust in the third party libs not to do something surprising or to not do something expected, so I write tests with examples of stuff I want to do to ensure it's going to work.

Reply:
But I would argue that all this additional code has maintenance and runtime costs, and that if we decide that we want to write test coverage for library dependencies, then we have to cover dozens of third party libraries in our dependency tree, which nobody would ever take on. I am sure that the json-schema library has coverage on array types and that we gain little to no benefit in the cost of keeping and maintaining that test code.

Reply:
I don't suggest that we write tests for every single third-party library that we use, but I do think it is useful to have some jsonschema tests that provide specific use case examples where the existing JSONSchema documentation is lacking or where there's some unexpected behaviour to be elucidated. I frequently look through tests to find usage examples if the docs aren't up to snuff. I've added a number of JSON validation tests specifically because the existing documentation of some feature or other was poor, I couldn't find any examples, and the best way to investigate what was going on was to create test cases and check that they worked.

In the specific case here, I've been trying to find out why array defaults weren't being filled in, whether it was a bug in `python-jsonschema`, some weird `pyyaml` thing, a problem with the schema itself, etc. It turns out that it's a shortcoming of the way that defaults are implemented -- not directly a jsonschema problem, but one due to our copying the simple default setting method in the jsonschema docs. I've written a test that directly illustrates the problem, as well as similar cases that do work.